src/main/java/frc/robot/subsystems/vision/VisionSubsystem.java

```diff
@@ -1,175 +1,306 @@
-   */
-  public Pose2d getPoseFromAprilTags() {
-    if (canSeeAprilTags()) {
-      // This is the MT2 Code, its not
-      // LimelightHelpers.SetRobotOrientation(VisionConstants.SHOOTER_LIMELIGHT_NAME, currentRobotHeadingDegrees, 0, 0, 0, 0, 0);
-      // return LimelightHelpers.getBotPoseEstimate_wpiBlue_MegaTag2(currentlyUsedLimelight).pose;
-      Pose2d botPose = LimelightHelpers.getBotPose2d(currentlyUsedLimelight);
-      // The origin of botpose is at the center of the field
-      double robotX = botPose.getX() + FieldConstants.FIELD_LENGTH_METERS / 2.0;
-      double robotY = botPose.getY() + FieldConstants.FIELD_WIDTH_METERS / 2.0;
-      Rotation2d robotRotation = botPose.getRotation();
-      lastSeenPose = new Pose2d(robotX, robotY, robotRotation);
-      return new Pose2d(robotX, robotY, robotRotation);
+
+    // double distanceToAprilTags = getLimelightAprilTagDistance(limelightNumber);
+
+    // if (isLargeDiscrepancyBetweenMegaTag1And2(limelightNumber)
+    //     && distanceToAprilTags < VisionConstants.MEGA_TAG_2_DISTANCE_THRESHOLD) {
+    //   limelightEstimates[limelightNumber] = getMegaTag1PoseEstimate(limelightNumber);
+    // } else
+    if (headingRateDegreesPerSecond < VisionConstants.MEGA_TAG_2_MAX_HEADING_RATE) {
+      LimelightHelpers.SetRobotOrientation(
+          getLimelightName(limelightNumber), headingDegrees, 0, 0, 0, 0, 0);
+      limelightEstimates[limelightNumber] = getMegaTag2PoseEstimate(limelightNumber);
     } else {
-      return new Pose2d();
+      limelightEstimates[limelightNumber] = getMegaTag1PoseEstimate(limelightNumber);
     }
   }
 
   /**
-   * Returns the distance in meters from the limelight(s) to the closest 
-   * april tag that they can see.
+   * Checks if there is a large discrepancy between the MegaTag1 and MegaTag2 estimates.
+   *
+   * @param limelightNumber the number of the limelight
+   * @return true if the discrepancy is larger than the defined threshold, false otherwise
    */
-  public double getDistanceFromClosestAprilTag() {
-    if (canSeeAprilTags()) {
-      int closestAprilTagID = (int) LimelightHelpers.getFiducialID(currentlyUsedLimelight);
-      double distance = getLimelightAprilTagDistance(closestAprilTagID);
-      // SmartDashboard.putNumber("distance from apriltag", distance);
-      return distance;
-    }
-    
-    // To be safe returns a big distance from the april tags if it can't see any
-    return Double.MAX_VALUE;
+  public boolean isLargeDiscrepancyBetweenMegaTag1And2(int limelightNumber) {
+    PoseEstimate megaTag1Estimate = getMegaTag1PoseEstimate(limelightNumber);
+    PoseEstimate megaTag2Estimate = getMegaTag2PoseEstimate(limelightNumber);
+
+    // Extract the positions of the two poses
+    Translation2d mt1 = megaTag1Estimate.pose.getTranslation();
+    Translation2d mt2 = megaTag2Estimate.pose.getTranslation();
+
+    // Calculate the Euclidean distance between the two positions
+    double euclideanDistance =
```

Comment:
while this works, translation2d has a [built in function](https://github.wpilib.org/allwpilib/docs/release/java/edu/wpi/first/math/geometry/Translation2d.html#getDistance(edu.wpi.first.math.geometry.Translation2d)) to get distance between 2 translations.

Also, what are your thoughts on actually using the translation3ds? That way, if for some reason the mt2 thinks we're 6 feet in the air, it could help filter it out.

Reply:
lol im blind. 

For the translation3ds, mt1 seems like it’d be more likely to be [flying around](https://docs.limelightvision.io/docs/docs-limelight/pipeline-apriltag/apriltag-robot-localization-megatag2) in the air while mt2 stays on the ground. mt2 would probably have a jank translation or smth due to not being zeroed correctly or smth.  

Yeah so I’d be fine with adding them.

Reply:
cool